### ISSUE TITLE
iOS System.NotImplementedException when using Callee Proxy

### ISSUE THREAD

**Issue by jaimewyant** [Code-Sharp/WampSharp#257]
My code looks like this:

```
            var factory = new WampChannelFactory();

            var authenticator = new TicketAuthenticator("myemailaddress", "mypassword");

            var channel = factory.ConnectToRealm("seasources")
                                 .WebSocketTransport(new Uri("ws://localhost:8080/ws"))
                                 .JsonSerialization()
                                 .Authenticator(authenticator)
                                 .Build();
            await channel.Open();

            var proxy = (IUser) channel.RealmProxy.Services.GetCalleeProxy<IUser>();

// My CalleeProxy interface

    public interface IUser
    {
        [WampProcedure("net.seasources.user.dictionary")]
        Task<JObject> GetUserDictionary();
    }
```
The exception occurs when I call GetCalleeProxy.  Below is the stack trace.  The code works on UWP / Android.

> System.NotImplementedException: The method or operation is not implemented.
>   at System.Reflection.DispatchProxy.Create[T,TProxy] () [0x00000] in <c0072a052bfd411ba72e00847966c95a>:0
>   at WampSharp.V2.CalleeProxy.WampCalleeClientProxyFactory.GetProxy[TProxy] (WampSharp.V2.ICalleeProxyInterceptor interceptor) [0x00000] in D:\a\1\s\src\net45\WampSharp\WAMP2\V2\System.Reflection.DispatchProxy\WampCalleeClientProxyFactory.cs:20
>   at WampSharp.V2.WampRealmProxyServiceProvider.GetCalleeProxy[TProxy] () [0x00000] in D:\a\1\s\src\net45\WampSharp\WAMP2\V2\Api\Client\WampRealmProxyServiceProvider.cs:83
>   at WampTest.Views.ItemsPage+<Handle_Clicked>d__4.MoveNext () [0x000ce] in /Users/jaimewyant/Projects/WampTest/WampTest/Views/ItemsPage.xaml.cs:65
>   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.12.0.4/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/exceptionservices/exceptionservicescommon.cs:152
>   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__6_0 (System.Object state) [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.12.0.4/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/AsyncMethodBuilder.cs:1018
>   at UIKit.UIKitSynchronizationContext+<>c__DisplayClass1_0.<Post>b__0 () [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.12.0.4/src/Xamarin.iOS/UIKit/UIKitSynchronizationContext.cs:24
>   at Foundation.NSAsyncActionDispatcher.Apply () [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.12.0.4/src/Xamarin.iOS/Foundation/NSAction.cs:125
>   at at (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
>   at UIKit.UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) [0x00005] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.12.0.4/src/Xamarin.iOS/UIKit/UIApplication.cs:79
>   at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x0002c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.12.0.4/src/Xamarin.iOS/UIKit/UIApplication.cs:63
>   at WampTest.iOS.Application.Main (System.String[] args) [0x00001] in /Users/jaimewyant/Projects/WampTest/WampTest.iOS/Main.cs:17
>

**Comment by darkl** [comment]
I guess that iOS has no DispatchProxy support. Too bad. iOS doesn't support System.Reflection.Emit either.

The only alternative I have is to generate classes that implement your interfaces at compile time. It is less convenient, but it will let you work with callee proxies also on Xamarin.iOS.

From now and on (v18.7.1), using the CalleeProxy api on iOS with an interface, will result with an exception containing an implementation of the interface as a property of the inner exception. This class can be passed safely as a generic parameter of GetCalleeProxy.

You can also generate these proxies manually by using the CalleeProxyCodeGenerator class, located in WampSharp.CodeGenerator.

Elad

**Comment by jaimewyant** [comment]
Thanks Elad.  I'll whip something together tonight that generates the implementation files.

Once I have a working solution, I'll post it up and close this issue.  Thanks.

**Comment by jaimewyant** [comment]
I have created a small program for generating these implementation files.  You can grab it here:

https://github.com/jaimewyant/WampSharpProxyMaker

If anyone has a better way of doing this, please let me know.  Thanks for your help Elad.

**Comment by darkl** [comment]
Cool.

No need to generate the interceptor interface. Just pass the generated
proxy class type to the GetCalleeProxy<> method. It automatically forwards
the interceptor received in that method to the generated class.

Elad

**Comment by jaimewyant** [comment]
@darkl is the updated GetCalleeProxy<> method supposed to work the same on all platforms.

I tried using:

```c#
IArgumentsService proxy = channel.RealmProxy.Services.GetCalleeProxy<ArgumentsServiceProxy>();
```

on UWP and I end up with an ArgumentException.

> {System.ArgumentException: The type 'UserProxy' must be an interface, not a class.
> Parameter name: T

**Comment by darkl** [comment]
Sorry, I'll fix that.

Elad

**Comment by darkl** [comment]
Ok, should be fixed in v18.7.2.

Elad

**Comment by jaimewyant** [comment]
Indeed it is, thanks.
